<a href="https://colab.research.google.com/github/Zach-Hanson3/Food_Sales_Prediction/blob/main/Food_Sales_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center> Food Sales Predictions </center>
#### <center> Zach Hanson </center>

## Loading Data

In [295]:
#Importing pandas
import pandas as pd

#Mounting drive 
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [296]:
#Saving file path
filename = '/content/drive/MyDrive/Coding Dojo/sales_predictions.csv'

#Creating data frame using file path
df = pd.read_csv(filename)

#Creating a copy that we can safely alter without worrying about losing information
sales_df = df.copy()

sales_df.info()
sales_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

## Data Cleaning

In [297]:
#Start by finding the shape of this data frame
sales_df.shape

(8523, 12)

We can see there is 8523 entries (rows) and 12 different variables (columns).

In [298]:
#Find the data types of each of the different variables
sales_df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

###Data Frame Cleaning


In [299]:
#Checking for duplicate rows
sales_df.duplicated().sum()

0

Sum of duplicated rows was 0, so we do not have any duplicated rows in our data.

In [300]:
#Checking to see if there is any missing values
sales_df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

There is 1463 values missing from Item_Weight and 2410 values missing from Outlet_Size.

In [301]:
#Removing the column of "Item_Weight"
sales_df = sales_df.drop(columns=['Item_Weight'])

#Data frame head to make sure column was dropped successfully
sales_df.head()

Item_Identifier Item_Fat_Content  Item_Visibility              Item_Type  \
0           FDA15          Low Fat         0.016047                  Dairy   
1           DRC01          Regular         0.019278            Soft Drinks   
2           FDN15          Low Fat         0.016760                   Meat   
3           FDX07          Regular         0.000000  Fruits and Vegetables   
4           NCD19          Low Fat         0.000000              Household   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  249.8092            OUT049                       1999      Medium   
1   48.2692            OUT018                       2009      Medium   
2  141.6180            OUT049                       1999      Medium   
3  182.0950            OUT010                       1998         NaN   
4   53.8614            OUT013                       1987        High   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0               Tier 1  Supermarket Type1          3735.1380  
1               Tier 3  Supermarket Type2           443.4228  
2               Tier 1  Supermarket Type1          2097.2700  
3               Tier 3      Grocery Store           732.3800  
4               Tier 3  Supermarket Type1           994.7052

I decided to remove the "Item_Weight" column because we were missing nearly 1500 entries in this column. While this alone would not warrant the removal of this variable, it also seems extremely unlikely that an item's weight would have any impact on it's sales.

In [302]:
#Replacing all the missing 'Outlet_Size' values with "Unknown"
sales_df = sales_df.fillna("Unknown")

In this case I chose to keep the "Outlet_Size" column because I do believe it might be valuable in determining the sales of an item. I replaced all the missing values with "Unknown".

In [303]:
#Double checking to make sure we removed all missing values
sales_df.isna().sum()

Item_Identifier              0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

- All the missing data has been removed

In [304]:
#Previewing data again with "unknown filled in for remaining missing values"
sales_df.head()

Item_Identifier Item_Fat_Content  Item_Visibility              Item_Type  \
0           FDA15          Low Fat         0.016047                  Dairy   
1           DRC01          Regular         0.019278            Soft Drinks   
2           FDN15          Low Fat         0.016760                   Meat   
3           FDX07          Regular         0.000000  Fruits and Vegetables   
4           NCD19          Low Fat         0.000000              Household   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  249.8092            OUT049                       1999      Medium   
1   48.2692            OUT018                       2009      Medium   
2  141.6180            OUT049                       1999      Medium   
3  182.0950            OUT010                       1998     Unknown   
4   53.8614            OUT013                       1987        High   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0               Tier 1  Supermarket Type1          3735.1380  
1               Tier 3  Supermarket Type2           443.4228  
2               Tier 1  Supermarket Type1          2097.2700  
3               Tier 3      Grocery Store           732.3800  
4               Tier 3  Supermarket Type1           994.7052

In [305]:
#Checking inconsistent data in Item_Fat_Content category
sales_df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

- We can see there is also values for "LF", "reg", and "low fat"

In [306]:
#Updating this inconsistent data so everything is the same
sales_df = sales_df.replace(['LF', 'low fat', 'reg'], ['Low Fat', 'Low Fat', 'Regular'])

#Checking to make sure everything got replaced
sales_df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

- Updated 'LF' and 'low fat' to 'Low Fat'
- Updated 'reg' to 'Regular'

In [307]:
#Checking 'Item_Type' column for inconsistencies
sales_df['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

- No inconsistencies noticed

In [308]:
#Checking 'Outlet_Size' for inconsistencies
sales_df['Outlet_Size'].value_counts()

Medium     2793
Unknown    2410
Small      2388
High        932
Name: Outlet_Size, dtype: int64

- No inconsistencies noticed

In [309]:
#Checking 'Outlet_Location_Type' for inconsistencies
sales_df['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

- No inconsistencies noticed

In [310]:
#Checking 'Outlet_Type' for inconsistencies
sales_df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

- No inconsistencies noticed

###Summary Statistics

In [311]:
#Preview of data for easy coding
sales_df.head()

Item_Identifier Item_Fat_Content  Item_Visibility              Item_Type  \
0           FDA15          Low Fat         0.016047                  Dairy   
1           DRC01          Regular         0.019278            Soft Drinks   
2           FDN15          Low Fat         0.016760                   Meat   
3           FDX07          Regular         0.000000  Fruits and Vegetables   
4           NCD19          Low Fat         0.000000              Household   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  249.8092            OUT049                       1999      Medium   
1   48.2692            OUT018                       2009      Medium   
2  141.6180            OUT049                       1999      Medium   
3  182.0950            OUT010                       1998     Unknown   
4   53.8614            OUT013                       1987        High   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0               Tier 1  Supermarket Type1          3735.1380  
1               Tier 3  Supermarket Type2           443.4228  
2               Tier 1  Supermarket Type1          2097.2700  
3               Tier 3      Grocery Store           732.3800  
4               Tier 3  Supermarket Type1           994.7052

In [312]:
#Creating dictionary with each numerical column, pointing to a list that will contain mean, median, and mode values for each column
sum_stat = {'Item_Visibility': {'Mean': '0', 'Median': '0', 'Mode': '0'}, 
            'Item_MRP': {'Mean': '0', 'Median': '0', 'Mode': '0'}, 
            'Outlet_Establishment_Year': {'Mean': '0', 'Median': '0', 'Mode': '0'}, 
            'Item_Outlet_Sales': {'Mean': '0', 'Median': '0', 'Mode': '0'}}

In [313]:
#Item_Visibility summary statistics 
sum_stat['Item_Visibility']['Mean'] = sales_df['Item_Visibility'].mean().round(5)
sum_stat['Item_Visibility']['Median'] = sales_df['Item_Visibility'].median().round(5)
sum_stat['Item_Visibility']['Mode'] = sales_df['Item_Visibility'].mode()

In [314]:
#Item_MRP summary statistics 
sum_stat['Item_MRP']['Mean'] = sales_df['Item_MRP'].mean().round(5)
sum_stat['Item_MRP']['Median'] = sales_df['Item_MRP'].median().round(5)
sum_stat['Item_MRP']['Mode'] = sales_df['Item_MRP'].mode()

In [315]:
#Outlet_Establishment_Year summary statistics 
sum_stat['Outlet_Establishment_Year']['Mean'] = sales_df['Outlet_Establishment_Year'].mean().round(5)
sum_stat['Outlet_Establishment_Year']['Median'] = sales_df['Outlet_Establishment_Year'].median().round(5)
sum_stat['Outlet_Establishment_Year']['Mode'] = sales_df['Outlet_Establishment_Year'].mode()

In [316]:
#Item_Outlet_Sales summary statistics 
sum_stat['Item_Outlet_Sales']['Mean'] = sales_df['Item_Outlet_Sales'].mean().round(5)
sum_stat['Item_Outlet_Sales']['Median'] = sales_df['Item_Outlet_Sales'].median().round(5)
sum_stat['Item_Outlet_Sales']['Mode'] = sales_df['Item_Outlet_Sales'].mode()

- Updating the summary statistics dictionary with all the actual values

In [317]:
#Printing all the summary statistics for each numerical column
for key in sum_stat:
  print(f"Summary Statistics for {key}:")
  print(sum_stat[key])
  print("")

Summary Statistics for Item_Visibility:
{'Mean': 0.06613, 'Median': 0.05393, 'Mode': 0    0.0
dtype: float64}

Summary Statistics for Item_MRP:
{'Mean': 140.99278, 'Median': 143.0128, 'Mode': 0    172.0422
dtype: float64}

Summary Statistics for Outlet_Establishment_Year:
{'Mean': 1997.83187, 'Median': 1999.0, 'Mode': 0    1985
dtype: int64}

Summary Statistics for Item_Outlet_Sales:
{'Mean': 2181.28891, 'Median': 1794.331, 'Mode': 0    958.752
dtype: float64}



Question: I recognize after printing the 'Mode' for summary statistics it shows a zero followed by the actual value. However, I cannot figure out why; so any insight here would be greatly appreciated :)

## Exploratory Visuals

In [11]:
#Exploratory Visuals stuff here

## Explanatory Visuals

In [12]:
#Explanatory Visuals stuff here